File: ruby/robot-name/robot_name.rb
require "set"

class Robot
  attr_reader :name
  @@used_names = Set.new

  def initialize
    set_name
  end

  def set_name
    @name = get_unique_name
    @@used_names << @name
  end

   def reset
    @@used_names.delete(@name)
    set_name
  end
  
  def get_unique_name
    loop do
      candidate = generate_random_name
      return candidate unless @@used_names.include? candidate
    end
  end

  def generate_random_name
    name = []
    name << ("A".."Z").to_a.sample(2)
    name << (0..9).to_a.sample(3)
    name.join
  end

  private :get_unique_name, :generate_random_name

end

module BookKeeping
  VERSION=2
end

File: python/robot-name/robot_name.py
import string
import random

class Robot(object):
    used_names = []

    def __init__(self):
        self.name = self.create_random_name()

    def create_random_name(self):
        numbers = ''.join(random.choices(string.digits, k=3))
        letters = ''.join(random.choices(string.ascii_uppercase, k=2))
        name = letters + numbers

        if name not in Robot.used_names:
            Robot.used_names.append(name)
            return name
        return self.create_random_name()


    def reset(self):
        new_name = self.create_random_name()

        if self.name in Robot.used_names: Robot.used_names.remove(self.name)
        self.name = new_name
